### PR TITLE
AGT-61 Added graphql http client

### PR DESCRIPTION
**Pull request by pec1985**
**JIRA:** https://pinpt-hq.atlassian.net/browse/AGT-61

### CODE REVIEW

**Comment by rblalock** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- graphql/client.go  10
- graphql/client_test.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Pinpoint/go-common/pullRequest?prid=4915263&bid=16268747)

File: graphql/client.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	return &client{
+		headers: headers,
+		url:     u.String(),
+	}, nil
+}
+
+func defaultHeaders(customerID string, userID string, apikey string) (map[string]string, error) {
+	headers := map[string]string{}
+	headers["x-api-key"] = apikey
+	if customerID != "" {
```
Comment (jhaynie) [inline graphql/client.go]: shouldnt we check for apikey != “” here too since you use it?

File: graphql/client.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+		Variables Variables `json:"variables"`
+		Query     string    `json:"query"`
+	}{variables, query}
+
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, c.url, bytes.NewBuffer(data))
```
Comment (jhaynie) [inline graphql/client.go]: need to use the api package yo construct so we can get benefit of tls cert checking

File: graphql/client.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	if err != nil {
+		return err
+	}
+	req.Header.Add("x-trace-id", hash.Values(9999*rand.Float64(), time.Now().Unix()))
```
Comment (jhaynie) [inline graphql/client.go]: remove trace header

File: graphql/client.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	for k, v := range c.headers {
+		req.Header.Add(k, v)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (jhaynie) [inline graphql/client.go]: probably should check status code too

File: graphql/client.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+// Mutate makes _query_ http request to the GraphQL server
+func (c *client) Query(query string, variables Variables, out interface{}) error {
+	return c.do(query, variables, out)
+}
+
+func (c *client) do(query string, variables Variables, out interface{}) error {
+
+	payload := struct {
```
Comment (jhaynie) [inline graphql/client.go]: like our other apis we might want to add automatic retry

File: graphql/client_test.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import (
+	"github.com/pinpt/go-common/hash"
+	"github.com/stretchr/testify/assert"
+	"math/rand"
+	"os"
+	"testing"
+)
+
+func TestClient(t *testing.T) {
+	if os.Getenv("CI") != "" {
```
Comment (jhaynie) [inline graphql/client_test.go]: we should also write unit tests that use httptest and mock out graphql endpoint

File: graphql/client.go (hunk 5 above)
Comment (pec1985) [reply graphql/client.go]: the api.NewHTTPAPIClientDefault should take care of this

File: graphql/client_test.go (hunk 6 above)
Comment (pec1985) [reply graphql/client_test.go]: let's discuss this offline

**Comment by rblalock** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- graphql/client.go  12
- graphql/client_test.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Pinpoint/go-common/pullRequest?prid=4915263&bid=16268747)

File: graphql/client_test.go (hunk 6 above)
Comment (pec1985) [reply graphql/client_test.go]: I wasn't familiar with the `httptest` package, I will modify the unit tests in a bit. Thanks for looking out

File: graphql/client.go (hunk 5 above)
Comment (jhaynie) [reply graphql/client.go]: it should also allow us to add support for authorization so an end customer with an api key can use not just an internal api key for service-to-service usage.

**Comment by rblalock** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- graphql/client.go  12
- graphql/client_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Pinpoint/go-common/pullRequest?prid=4915263&bid=16268747)

**Comment by rblalock** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- graphql/client.go  13
- graphql/client_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Pinpoint/go-common/pullRequest?prid=4915263&bid=16268747)

**Pull request merged**